### ISSUE TITLE
invite user functionality for admins

### ISSUE THREAD

**Issue by eplebel** [ScienceCommons/curate_science#14]
admins need to be able to  invite new users so that they can create their own author page. 

perhaps for now, clicking "Invite new user" (from user menu bar, see mockup in #13)  can open a modal window with email address and name text fields and a "Send Invitation" button. this will send an email to the user with a signup invitation link. 

when clicked, the user will go to signup page, which will have their email and name pre-populated, and password field (which user will have to copy-and-paste from the email?). 

clicking "Sign up" will then create the new user account, log-in the user, and bring them to the homepage in a logged-in state (where they can then click "Create author page".

>Alex comments

_E-mail address is the username, but I do need to implement a feature for the ability for admins to send invites. Looks like there are existing packages for this, so it shouldn't be much work. And since only admins will have invite permissions, there's no need for an invite quota. But if you decide you want to give authors the ability to send invites with a quota, let me know._

> Etienne response

_ok sounds good. and indeed authors should have an invite quota, which initially will be set to 0. only once the author page is tested and improved based on a small group of users (~15-20), will we want to give 2-3 invites to each user._ ***but i'm not sure if you want to implement this now, or just add it later.***

**Comment by alexkyllo** [comment]
@onejgordon what will be the URL path of the sign up page that the user lands on when they click the link in the invite e-mail? I'm adding a REST API endpoint at /api/accounts/create/. 
@eplebel do you have a spec or mock-up for how you want the signup page to look? I'm assuming @onejgordon will be implementing that as part of the frontend UI.

**Comment by onejgordon** [comment]
I hope this can be customized, but for now let's use /app/signup

**Comment by alexkyllo** [comment]
Yeah it'll be a one-liner in the settings.py file.

**Comment by alexkyllo** [comment]
@eplebel what e-mail account do you want the invite e-mails to be sent out from? I will need the password in order to get it working, so I suggest setting up a new, unmonitored inbox at curatescience.org just for this purpose.

Do you have a preference for what you want the e-mail subject line and body to say?

The default message looks like this when I generate it locally:

Subject: [example.com] Invitation to join example.com
From: webmaster@localhost
To: alex.kyllo@gmail.com
Date: Wed, 20 Feb 2019 07:10:10 -0000

Hello,

You (alex.kyllo@gmail.com) have been invited to join example.com

If you'd like to join, please go to http://localhost:8000/invitations/accept-invite/geq15u2wgrootffkgasd8vvq8xcrzucxkhtul80jkv7gmxerfilyydtwgtozmdxb

**Comment by eplebel** [comment]
would it be possible to use "curatescience@gmail.com" (which i monitor)? this is a free non-domain-name-specific email address (but i assume we can change it later once we re-activate "@curatescience.org" email addresses, which i de-activated in 2016 given we weren't using them). i can send you the password through google drive.

for the message, perhaps change it slightly to:

_Hello [name],

You (email address) have been invited to join Curate Science, an exciting new platform to label and link the transparency of your research.

To join, please go to (link)

Regards,
Curate Science team_

**Comment by alexkyllo** [comment]
That works, I'll get it hooked up.

**Comment by onejgordon** [comment]
It looks like this invite flow doesn't actually involve a signup UI, is that right? Is one needed or will users simply click the invite link to perform account creation?

**Comment by alexkyllo** [comment]
It does need a simple signup UI because we're using basic auth, so the user has to at least set their password.

**Comment by onejgordon** [comment]
Sounds good. Though it's simple, we should probably create a wireframe of the UI to make sure we're on the same page, and to confirm the invite flow. @eplebel do you want to do this wireframe since you've done all the others, or would you like me to?

**Comment by alexkyllo** [comment]
If we dropped basic auth and switched to using some social OAuth system (Sign in with Google, Facebook, Microsoft, GitHub etc) then we could simply grant them access without asking them to submit a form.

**Comment by eplebel** [comment]
ya if it's simpler for now, we could *only* offer logging-in (authentication) via Google, Facebook, and Twitter given that these 3 options (most likely) covers > 95% of users. 

and then later we could offer a non-social login where a password needs to be set (for the rare user who doesn't have any of these).

**Comment by alexkyllo** [comment]
It's a tradeoff--in order to integrate with social OAuth2 providers I'll need to go register our app with each one of them, and Jeremy would still need to make a simple signup/login page for them to land on from the e-mail invite and click a button to select which social auth provider they want to use.

The only thing it would really save us is the effort of implementing password change/reset flow. Which is pretty simple on the Django side, there's a package I can use with just a little configuration.

**Comment by eplebel** [comment]
ok sounds good (we can add social auth login options a bit later)

**Comment by onejgordon** [comment]
@alexkyllo is there an API endpoint for sending user invites?

I've got a simple page for admins to handle this, but it's not hooked up to anything.

<img width="1295" alt="screen shot 2019-03-03 at 4 46 45 pm" src="https://user-images.githubusercontent.com/57584/53704705-f9cf0a80-3dd3-11e9-8804-7855e62ad7ea.png">

Could you also clarify the flow that the frontend needs to support for fulfilling invitations and creating the user account?

**Comment by alexkyllo** [comment]
There isn't yet but I will add one at /api/invitations/create/

Within the Django admin site it is currently possible to create invitations. Go to /admin/ and log in and click Invitations.

The email that they click on should land them at a sign up page, and then the sign up page should POST to /api/accounts/create.

I believe the URL from the email contains the sign up code as a parameter, and that param should be included in the request to /accounts/create/, I think that's how it prevents non-invited users from signing up. I will check and confirm this.

**Comment by alexkyllo** [comment]
By the way, since the login page is an HTML form, should the signup page be as well? It might be easier for me to implement it with a Django form than it is to go through the REST API.

**Comment by onejgordon** [comment]
> There isn't yet but I will add one at /api/invitations/create/

Great.

> easier for me to implement it with a Django form than it is to go through the REST API. I can redirect the user to /app/ after the sign up flow is complete.

Works for me. I'll style this and the login page with CSS only.

**Comment by alexkyllo** [comment]
@onejgordon , I've created an API endpoint at `/api/invitations/create/` that is only accessible to admins. The JSON body needs one key which is "email", a string containing the e-mail address of the person you're inviting.

Upon clicking the link in the invitation e-mail, the user will be redirected to `/accounts/signup/` with their e-mail pre-filled, and will need to enter a username (which can be same as e-mail) and password.

If someone tries to access /accounts/signup/ without an invite code, they will just see the following message:

> Sign Up Closed
>
> We are sorry, but the sign up is currently closed.

**Comment by eplebel** [comment]
> Upon clicking the link in the invitation e-mail, the user will be redirected to /accounts/signup/ with their e-mail pre-filled, and will need to enter a username (which can be same as e-mail) and password.

for user names , given that we'll eventually display user names (e.g., on user pages; to show who contributed/changed a piece of information), it's probably good to follow the same approach we're using for author page URL generation (https://github.com/ScienceCommons/curate_science/issues/12#issuecomment-465905323), except use periods ('.') instead of hyphens ('-').

that is, user names should be a "periodified" version of the *user's full name*, which should be automatically generated and displayed on the signup page (and be non-editable). for non-unique cases, an integer is automatically added (e.g., if publication-less Master's student "Martin S Smith" is invited, his username would be "martin.s.smith"; then if a published PhD student also called "Martin S Smith" is invited, his username would be "martin.s.smith2").

**Comment by alexkyllo** [comment]
I think I can do something like that, I could create a UserProfile object at the time the Invitation is created, and populate a username field with a generated slug based on the full name. Then hide it from the signup form.

It'd be more complicated to implement it retroactively once we have an existing user base, so it's probably better to do it now.

Just to let you know where I'm at, I've logged 15 hours so far, and I believe I've checked off all the items on my to-do list except for this username generation feature you describe above. Do you want me to go ahead with it?

**Comment by onejgordon** [comment]
I'm concerned about potential confusion related to the fact that a single
user may be associated with two slugified names, possibly the same and
possibly different (e.g. if jeremy.gordon was taken among authors but not
yet for users).

We could avoid this by avoiding a second unique id in username.

When ready to show editors in the UI could we not just use the user's full
name or even a non-unique alias and then link to a profile page?

**Comment by alexkyllo** [comment]
That's a good point @onejgordon, though I think it's what @eplebel was addressing with the periods instead of dashes.

Actually my personal preference would be to make the username always automatically be the same as the email address, and then have the UI just show the full name where appropriate.

If users don't get to choose their own usernames, then there's no point in having a username that's different from their email address.

**Comment by eplebel** [comment]
> make the username always automatically be the same as the email address, and then have the UI just show the full name (hyperlinked to their profile page) where appropriate.

the reason to want a unique user name (internally and displayed in the UI) is to allow for reputational dynamics learning to take place. one needs to be able to distinguish between user contributions so one can learn about the trustworthiness/accuracy of different users' contributions (e.g., i learn that jeremy.gordon's contributions are almost always correct *whereas* jeremy.gordon2's are often inaccurate). hence, why we can't just use/show non-unique full names as user names (and we also can't use/show email addresses, because these need to remain private). 

it's true that confusion may arise in rare cases where a single user is associated with two different slugified names. i guess a potential solution could be to force the *same* slugified names for both author URL and user name (except use periods for user names because i don't think hyphens can be used for user names right?). 

so when an author page *or* user is created, the system automatically determines the next unique slug for both author URL *and* user names (e.g., if two "Jeremy Gordon's" signed up before you, then your author URL would be "jeremy-gordon-3" and your user name would be "jeremy.gordon3" [or "jeremy.gordon.3"]).  

thoughts?

**Comment by eplebel** [comment]
i guess we could, but we'd want to constrain it to slugified full names anyway (again for reputational dynamics; we can't have cryptic pseudonyms or silly stuff), so they really can only change/choose the integer at the end that makes their user name/author page URL unique. 

at that point, i don't think it's worth it because you need to do all this extra work just to allow people to choose between "jeremy-gordon-3" and "jeremy-gordon-81" or something... right?

**Comment by onejgordon** [comment]
If we're working with something like a real name policy (where users are expected to enter their actual names in their profile), do we need a concept of a username at all? I don't see what benefit we're getting from it. 

What if the only universally unique identifier for users is the db-generated integer, and users or admins can visit users profiles (app/user/4931) via a search by name/email interface. I see why we want nice looking slugified URLs for author pages, but I'm not seeing why this is necessary for user pages, especially if users have no control of it anyway.

**Comment by onejgordon** [comment]
@alexkyllo I'm getting `POST http://localhost:8000/api/invitations/create/ 415 (Unsupported Media Type)`. I suspect this may be a permissions problem, by the test user I'm using is staff and superuser.

**Comment by alexkyllo** [comment]
I just tried this and it worked, so I don't think I can reproduce the 415 error you're getting:

```
curl -X POST http://localhost:8000/api/invitations/create/ 
-H "Content-Type: application/json" 
-d '{"email": "my.address@gmail.com"}' 
-u adminuser:adminpassword
```

**Comment by onejgordon** [comment]
Thanks, problem was on my side.

**Comment by eplebel** [comment]
> If we're working with something like a real name policy (where users are expected to enter their actual names in their profile), do we need a concept of a username at all? I don't see what benefit we're getting from it. 

as i said in my comment yesterday: the reason we need a unique user name (displayed in the UI) is to allow for reputational dynamics learning to take place. one needs to be able to distinguish between user contributions so that one can learn about the trustworthiness/accuracy of different users' contributions (e.g., i need to be able to learn that jeremy-gordon's contributions are almost always correct whereas jeremy-gordon-2's are often inaccurate). hence, why we can't just show non-unique full names as user names (because there will be no way to distinguish the edits of "Jeremy Gordon" [jeremy-gordon] from "Jeremy Gordon" [jeremy-gordon-2]). 

hopefully that makes sense.

**Comment by alexkyllo** [comment]
@onejgordon I noticed you removed the name field from the invitation form. You can add it back, since we'll need it to create the username slug.

I just pushed a change to staging to enhance the /api/invitations/create/ endpoint so that you can:
```json
POST /api/invitations/create/ HTTP/1.1
Content-Type: application/json
{
	"email": "my.address@gmail.com",
	"userprofile": {
		"name": "Alex Kyllo"
	}
}
```
And it will create an Invitation and a UserProfile like this:

```json
HTTP 201 Created
Allow: OPTIONS, POST, GET
Content-Type: application/json
Vary: Accept

{
    "email": "my.address@gmail.com",
    "userprofile": {
        "curation_contributions": null,
        "browsing_history": null,
        "tracked_content": null,
        "recommended_content": null,
        "account_settings": null,
        "research_interests": null,
        "name": "Alex Kyllo",
        "slug": "alex-kyllo",
        "user": null,
        "invite": null
    }
}
```
Now I just need to figure out how to ensure that upon signup, the User account gets automatically linked to this UserProfile.

**Comment by onejgordon** [comment]
I've updated on staging but am getting 500s on staging when trying to invite. 

https://console.cloud.google.com/errors/CK3TsPHm6-23bQ?time=P1D&project=curatescience-staging&authuser=1

**Comment by alexkyllo** [comment]
That error is saying there's a table missing from the database, which means there's a migration that hasn't been run on staging. I thought I ran it though, I'll double check this evening.

**Comment by alexkyllo** [comment]
After running `python manage.py migrate` on staging the error went away.

Signup form no longer has a username field on it. I added an override for the save_user method called by the SignupForm, which will take the slug created at the time the Invitation was created, and fill it into the user's username field.

In the future, if we turn off the invitation system, then we'll need to ask the user to provide their own full name at the time of signup, to generate the slug and username.

**Comment by eplebel** [comment]
user names don't seem to match the author page URL which is needed as per my last comment on this thread (https://github.com/ScienceCommons/curate_science/issues/14#issuecomment-472316060)

![Untitled111111](https://user-images.githubusercontent.com/4512335/54529440-2f512780-4980-11e9-8cea-a7f677832ecc.png)

**Comment by alexkyllo** [comment]
Did you create multiple authors named "Steve Label" before creating one as the user "Steve Label"?

There's currently nothing forcing the user name and slug and author name and slug fields to be synchronized, so the two slugs being the same depends on the two names being input the same way with no duplicates in the system.

If it's important that they always match exactly, then I can add some code to overwrite one with the other, depending on which is created first. We should also make the UI so that users aren't even prompted to reenter their name when they create their author page, it just gets copied over from their User data.

**Comment by alexkyllo** [comment]
Also I think we need to more clearly specify the flow where an admin creates an author page and then a user can either "claim" it or create their own. 

If there are two Steve Labels and you create author profiles for them, one will be steve-label and the other steve-label-2. Then if you invite them in reverse order, one will have username steve-label and author slug steve-label-2, and vice versa. I'm sure there are other edge cases here as well...

**Comment by eplebel** [comment]
the following 2 requirements are necessary for the crowdsourcing platform to work and grow:
1. user names (displayed publicly and to the logged-in user themselves) need to be unique for [reputational dynamics to work](https://github.com/ScienceCommons/curate_science/issues/14#issuecomment-472316060) 
2. admins need to be able to create author pages (so that we can lure in prominent transparency mavericks by first [partially] curating their author pages *and then inviting* them to the platform).

fulfilling these requirements is definitely tricky, but i think can be accomplished by implementing 2 changes:
1. forcing user name and author name slugs to be the same.
2. allowing admins to invite users where an author page already exists for that user

i believe these 2 additions cover all of the (edge) cases, including:
1. "Steve Smith" signs up and doesn't create an author page (user name and author page slugs are "steve-smith" (which is used if they decide to create an author page later *or* if an admin creates an author page for them; more on this below)
2. a different "Steve Smith" signs up and creates an author page ("steve-smith-2")
3. an admin creates an author page for a different "Steve Smith" ("steve-smith-3"), and this user name is used to invite the user
4. an admin creates an author page for a pre-existing "Steve Smith" that doesn't yet have an author page (e.g., Steve Smith 1) and can enter this existing user name (after finding it via search/admin panel) on the Create Author page (already implemented)

actually, just before posting this comment, i realized there's probably an issue w/ the 3rd scenario, though i can't easily articulate it. perhaps this comment can nonetheless help us find a workable solution.

**Comment by eplebel** [comment]
> Did you create multiple authors named "Steve Label" before creating one as the user "Steve Label"?

oh and yes

**Comment by alexkyllo** [comment]
>>Did you create multiple authors named "Steve Label" before creating one as the user "Steve Label"?

>oh and yes

What would you expect to happen when you did this? 

To make the slugs match, you would have had to sign up as the user steve-label and then link your user profile to the existing steve-label author page (the first one you created as an admin), instead of creating a new author page.

**Comment by alexkyllo** [comment]
@eplebel here's a thought: What if creating an author page automatically created a linked user account, and vice versa? Then when you invite the user and they accept, their user account and author page are already linked and waiting for them, they would just need to set their password and start curating. This way, a given person's username and author slug would always match because they'd always be created at the same time, and they wouldn't need to decide between creating a new author profile vs searching for and linking one that you already created for them.

The only downside I can think of is that the users who don't need an author page would still have one, but in my mind that's OK because they could just not fill it out or link any articles to it.

**Comment by eplebel** [comment]
yes i think your proposed approach (of automatically/simultaneously creating/linking user and author page) is a good one. 

however, i still think it may not cover an edge case whereby an admin creates an author page, but then that person signs up *before* being invited by the admin (or somehow gets invited but doesn't click the invite link). for this edge case, the user would need to be able to claim a pre-existing author page and then any (non-redundant) articles on that pre-existing author page would be merged into their newest author page.

for example, admin creates author page for "steve-smith", but doesn't invite him. later, that same person signs up under "steve-smith-2". the user then realizes that the "steve-smith" author page is his, and claims the author page. all (non-redundant) articles from pre-existing "steve-smith" author page are merged into his "steve-smith-2" author page, and the "steve-smith" author page is deactivated (or could even re-direct to "steve-smith-2" author page; but any one searching for "Steve Smith" would only be displayed the correct activated "steve-smith-2" author page.

so in summary, it appears we need the following new functionality:
1. automatic/simultaneous creation/linking of user and author page (having same slug)
2. allow user to claim pre-existing author page
3. author page merge tool (merging all non-redundant articles of pre-existing author page into newest author-page)

well let me know what you think. 
this of course involves a non-trivial amount of work, but the good news is that we only need the stuff in 
# 1 for now and can implement # 2 and # 3 later (given that the 1st group of early beta testers won't be allowed to invite new users and non-invited users won't be able to sign-up until way later).
